src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,203 @@
+ */
+
+#include "postgres.h"
+#include "commands/copy.h"
+#include "catalog/namespace.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+#include "nodes/makefuncs.h"
+#include "safe_lib.h"
+#include <netinet/in.h> /* for htons */
+
+#include "distributed/transmit.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/local_executor.h"
+#include "distributed/local_multi_copy.h"
+#include "distributed/shard_utils.h"
+
+/*
+ * LOCAL_COPY_BUFFER_SIZE is buffer size for local copy.
+ * There will be one buffer for each local placement, therefore
+ * the maximum amount of memory that might be alocated is
+ * LOCAL_COPY_BUFFER_SIZE * #local_placement
+ */
+#define LOCAL_COPY_BUFFER_SIZE (1 * 512 * 1024)
+
+
+static int ReadFromLocalBufferCallback(void *outbuf, int minread, int maxread);
+static void AddSlotToBuffer(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest,
+							bool isBinary);
+
+static bool ShouldSendCopyNow(StringInfo buffer);
+static void DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId,
+						CopyStmt *copyStatement, bool isEndOfCopy);
+static bool ShouldAddBinaryHeaders(StringInfo buffer, bool isBinary);
+
+/*
+ * localCopyBuffer is used in copy callback to return the copied rows.
+ * The reason this is a global variable is that we cannot pass an additional
+ * argument to the copy callback.
+ */
+StringInfo localCopyBuffer;
+
+/*
+ * ProcessLocalCopy adds the given slot and does a local copy if
+ * this is the end of copy, or the buffer size exceeds the threshold.
+ */
+void
+ProcessLocalCopy(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest, int64 shardId,
+				 StringInfo buffer, bool isEndOfCopy)
+{
+	/*
+	 * Here we save the previous buffer, and put the local shard's buffer
+	 * into copyOutState. The motivation is to use the existing logic to
+	 * serialize a row slot into buffer.
+	 */
+	StringInfo previousBuffer = copyDest->copyOutState->fe_msgbuf;
+	copyDest->copyOutState->fe_msgbuf = buffer;
+
+	/* since we are doing a local copy, the following statements should use local execution to see the changes */
+	TransactionAccessedLocalPlacement = true;
+
+	bool isBinaryCopy = copyDest->copyOutState->binary;
+	AddSlotToBuffer(slot, copyDest, isBinaryCopy);
+
+	if (isEndOfCopy || ShouldSendCopyNow(buffer))
+	{
+		if (isBinaryCopy)
+		{
+			AppendCopyBinaryFooters(copyDest->copyOutState);
+		}
+
+		DoLocalCopy(buffer, copyDest->distributedRelationId, shardId,
+					copyDest->copyStatement, isEndOfCopy);
+	}
+	copyDest->copyOutState->fe_msgbuf = previousBuffer;
+}
+
+
+/*
+ * AddSlotToBuffer serializes the given slot and adds it to the buffer in copyDest.
+ * If the copy format is binary, it adds binary headers as well.
+ */
+static void
+AddSlotToBuffer(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest, bool isBinary)
+{
+	if (ShouldAddBinaryHeaders(copyDest->copyOutState->fe_msgbuf, isBinary))
```

Comment:
Suggestion: move this to `ProcessLocalCopy()`. It looks a bit weird to add footers there and headers here.